adjusted system prompt for agent, added capability to add parameter t…

JonasBaeumer opened the pull request:
## Description

Fixed the problem where the agentic quality check (e.g. applying filter such as publication year) to the returned results leads to empty results being returned because we apply the filter only to the 10 best papers being returned. 

What we now do is that get_best_papers has an additional param, which the agent can modify to increase the number of papers returned when it wants to apply a filter in the next step. This ensures that instead of the top 10 we will retrieve the top 100 papers and filter them leading to much more stable results. 

Retrieving more papers makes the get_best_papers tool slower but I think it is worth it. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [X] Bug fix

## How Has This Been Tested?

Run the code with tight filer constraints, such as from the year 2020 onwards to make sure you still see papers. Also check the logs to ensure there is a sufficient number of papers retrieved when filtering (e.g. 100) and filtered. 

## Reviewers

@Daniel @Aditi

daniel-emrani commented on requirements.txt:
```diff
@@ -6,7 +6,7 @@ chromadb==1.0.12
 langchain-openai==0.3.19
 langchain==0.3.25
-langgraph==0.5.0
+langgraph~=0.4.5
```
Please don't do this. We had compatibility issues with 0.4.5 so I'd like to keep it at 0.5.0 unless you have a reason not to. I assume this was done automatically and not intentionally by you anyways.

daniel-emrani commented:
Did not work for me. Tried with the prompt:
`Looking for papers like this BUT only from 2024 or later <Insert Attention Is All You Need paper>
`
And got no results, even though "get best papers" returned many results that would match the filter.


Here are the logs:
`apstone_web   | INFO:llm.tools.paper_handling_tools:Updating paper database successfully.
capstone_web   | ================================= Tool Message =================================
capstone_web   | Name: update_papers
capstone_web   | 
capstone_web   | Paper database has been updated with the latest papers & embeddings. There were no errors. Now you can rank the papers.
capstone_web   | INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
capstone_web   | ================================== Ai Message ==================================
capstone_web   | Tool Calls:
capstone_web   |  get_best_papers (call_njQSUqYXIf2WmKAaqLo5i9zG)
capstone_web   |  Call ID: call_njQSUqYXIf2WmKAaqLo5i9zG
capstone_web   |  Args:
capstone_web   |   user_profile: transformer architecture, attention mechanism neural networks, self-attention models, sequence transduction models, transformer neural networks, advances in transformer models, transformer variants 2024, attention-based deep learning 2024
capstone_web   |   num_candidates: 100
capstone_web   | INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings "HTTP/1.1 200 OK"
capstone_web   | INFO:paper_ranking.paper_ranker:Agent requested 100 candidates, which is different than the default of 10. This will allow for more filtering.
capstone_web   | INFO:httpx:HTTP Request: POST http://chromadb:8000/api/v2/tenants/default_tenant/databases/default_database/collections/29706cbd-becc-4df2-a6d6-1b1913b7635c/query "HTTP/1.1 200 OK"
capstone_web   | INFO:paper_ranking.paper_ranker:Paper metadata: {paper_metadata}
capstone_web   | ================================= Tool Message =================================
capstone_web   | Name: get_best_papers
capstone_web   |
---------
capstone_web   |   criteria_nl: Only include papers published in 2024 or later.
capstone_web   | INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
capstone_web   | INFO:llm.tools.paper_handling_tools:LLM-generated filter_spec: {'publication_date': {'op': '>=', 'value': 2024}}
capstone_web   | ================================= Tool Message =================================
capstone_web   | Name: filter_papers_by_nl_criteria
capstone_web   | 
capstone_web   | {"status": "success", "filters": {"publication_date": {"op": ">=", "value": 2024}}, "kept_count": 0, "kept_papers": [], "reasoning": "Applied 1 metric filter(s)"}
capstone_web   | INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
capstone_web   | ================================== Ai Message ==================================
capstone_web   | 
capstone_web   | {"papers": []}-`

JonasBaeumer commented:
@daniel-emrani here is also my response from slack: 

I added a one sentence explanation for each metric to the UI.
Second, I fixed the problem with the agent changing the format and trimming down the field earlier then necessary. Then I discovered another problem: Essentially the agent tends to forget parts of the papers when passing them down between the tools when they are part of the prompt. This is a known issue with LLM when they work with large data that is just pasted in as a prompt. Also here we will need stategraph which contains an external state that is not just a simple prompt but rather directly injected by the agent to the tools. For now I have reduced the number of top papers that are retrieved down to 50. This stabilized the agent. There can still be instances where the agent forgets part of the papers, but its seems much less likely for now. I didnt want to engineer a state file for now since that will come in stategraph so it would be duplicated work.
I reran you prompt 5 times now and I always gotten at least one response paper (4x 3 papers and 1x 1 paper). Please check and merge if you can. I hope this should fix the most urging issues for now.
